hedera-node/docs/design/services/smart-contract-service/contract-accounts-nonces.md

```diff
@@ -34,43 +25,64 @@ The following is a table with general use cases and behavior for Ethereum and He
+
+
+- We keep a `ContractId -> nonce` tree map inside `HederaWorldState`, it is updated on each call of `commit()` (using newly added method `trackContractNonces()`).
+- Method `trackContractNonces` in `HederaWorldState` follows the pattern of `trackNewlyCreatedAccounts`.
+    - Checks if an account is a new smart contract and externalizes its nonce.
+    - Checks if an existing smart contract's nonce is updated and externalizes it.
+- Added a `ContractId -> nonce` tree map inside `TransactionProcessingResult`.
+- Persists account contract nonces into state in `ContractCreateTransitionLogic` and `ContractCallTransitionLogic` using `setContractNonces` from `TransactionProcessingResult`.
+- Created new `ContractNonceInfo` submerkle class with two main entities - `contractId` and `nonce`
+- Added new method `serializableContractNoncesFrom` in `EvmFnResult` that builds `List<ContractNonceInfo>` (submerkle) from `Map<ContractID, Long>`
+- Added new verison `7` (`RELEASE_0400_VERSION`) and externalized logic for `serialize` and `deserialize` of `contractNonces` in `EvmFnResult`
+
+
+
+
+### Fix Storing Of Nonces Into State
+
+- Fix storing of nonces is done in the following places:
+  - [UpdateAccountTracker](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/UpdateAccountTracker.java)
+  - [UpdateAccountTrackerImpl](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/UpdateAccountTrackerImpl.java)
+  - [UpdateTrackingAccount](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/models/UpdateTrackingAccount.java)
+  - [HederaEvmEntityAccess](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/contracts/HederaEvmEntityAccess.java)
+  - [MutableEntityAccess](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/store/contracts/MutableEntityAccess.java)
+  - [WorldStateAccount](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-evm/src/main/java/com/hedera/node/app/service/evm/store/contracts/WorldStateAccount.java)
+  - [HederaLedger](https://github.com/hashgraph/hedera-services/blob/96a85f0e08f82582bbf25328d14ca90fc630c5ef/hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/mono/ledger/HederaLedger.java)
+  
+
+- Currently we are not storing contract account nonces into state and this needs a fix similar to `setBalance` in `UpdateTrackingAccount`.
```

Comment:
<!-- Suggestions are specific requests to improve the subject under review. It is assumed that we all want to do what's best, so these comments are never dismissed as “mere suggestions”, but are taken seriously.  -->
**suggestion:** ‏Let's add the above links into the proper places in this description.